testbed/tests/window/test_window.py

```diff
@@ -507,6 +689,130 @@ async def test_move_and_resize(second_window, second_window_probe):
+            (WindowState.FULLSCREEN, WindowState.FULLSCREEN),
+            # Switch from PRESENTATION:
+            (WindowState.PRESENTATION, WindowState.NORMAL),
+            (WindowState.PRESENTATION, WindowState.MINIMIZED),
+            (WindowState.PRESENTATION, WindowState.MAXIMIZED),
+            (WindowState.PRESENTATION, WindowState.FULLSCREEN),
+            (WindowState.PRESENTATION, WindowState.PRESENTATION),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "intermediate_states",
+        [
+            (
+                WindowState.MINIMIZED,
+                WindowState.FULLSCREEN,
+                WindowState.PRESENTATION,
+                WindowState.MAXIMIZED,
+                WindowState.MINIMIZED,
+                WindowState.FULLSCREEN,
+            ),
+            (
+                WindowState.FULLSCREEN,
+                WindowState.MAXIMIZED,
+                WindowState.MINIMIZED,
+                WindowState.PRESENTATION,
+                WindowState.FULLSCREEN,
+                WindowState.MINIMIZED,
+            ),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "second_window_class, second_window_kwargs",
+        [
+            (
+                toga.MainWindow,
+                dict(title="Secondary Window", position=(200, 150)),
+            )
+        ],
+    )
+    async def test_window_state_change_with_intermediate_states(
+        app,
+        app_probe,
+        initial_state,
+        final_state,
+        second_window,
+        second_window_probe,
+        intermediate_states,
+    ):
+        """Window state can be changed to another state while passing
+        through intermediate states with an expected OS delay."""
+        if (
+            WindowState.MINIMIZED in {initial_state, final_state}
+            and not second_window_probe.supports_minimize
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+        second_window.toolbar.add(app.cmd1)
+        second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
+        second_window.show()
+
+        # Add delay to ensure windows are visible after animation.
+        await second_window_probe.wait_for_window("Secondary window is visible")
+
+        assert second_window.state == WindowState.NORMAL
+
+        # Set to initial state
+        second_window.state = initial_state
+        # Add delay to ensure windows are visible after animation.
+        await second_window_probe.wait_for_window(
+            f"Secondary window is in {initial_state}",
+            minimize=True if initial_state == WindowState.MINIMIZED else False,
+            full_screen=True if initial_state == WindowState.FULLSCREEN else False,
+        )
+
+        assert second_window.state == initial_state
+
+        # Set to the intermediate states but don't wait for the OS delay.
+        for state in intermediate_states:
```

Comment:
Won't this be transitioning through intermediate states that the backend may not support?

Reply:
You are right, I have missed this case, it would be transitioning through unsupported window state like MINIMIZED on wayland. So, I have added a check to skip assigning the unsupported state. 

I know that the tests should minimize the amount of extra logic in them. But, the other two alternatives were to remove the unsupported state from the intermediate states set at the start of the test or to skip the testing altogether.

Reply:
I have reverted the change I had made, since assigning unsupported window states is a silent no-op. Trying to set the intermediate states ensures coverage on wayland.